msticpy/nbtools/timeline.py

```diff
@@ -0,0 +1,940 @@
+    group_by: str = kwargs.pop("group_by", None)
+    legend_column: str = kwargs.pop("legend_column", None)
+    legend_pos: str = kwargs.pop("legend", None)
+
+    # If legend_pos is outside the graph we need to create the legend
+    # seperately.
+    # We plot groups individually so that we can create an interactive legend.
+    legend_items = []
+    for _, group_id in group_count_df[group_by].items():
+        first_group_item = graph_df[graph_df[group_by] == group_id].iloc[0]
+        if legend_column:
+            legend_label = str(first_group_item[legend_column])
+        else:
+            legend_label = str(first_group_item[group_by])
+        inline_legend = legend_label if legend_pos == "inline" else None
+        row_source = ColumnDataSource(graph_df[graph_df[group_by] == group_id])
+        p_series = plot.diamond(
+            x=time_column,
+            y="y_index",
+            color="color",
+            alpha=0.5,
+            size=10,
+            source=row_source,
+            legend=inline_legend,
+        )
+        if not inline_legend:
+            legend_items.append((legend_label, [p_series]))
+
+    if legend_pos == "inline":
+        # Position the inline legend
+        plot.legend.location = "top_left"
+        plot.legend.click_policy = "hide"
+    elif legend_pos in ["left", "right"]:
+        # Create the legend box outside of the plot area
+        ext_legend = Legend(
+            items=legend_items,
+            location="center",
+            click_policy="hide",
+            label_text_font_size="8pt",
+        )
+        plot.add_layout(ext_legend, legend_pos)
+
+
+# pylint: enable=too-many-locals
+
+
+def _wrap_df_columns(data: pd.DataFrame, wrap_len: int = 50):
+    """Wrap any string columns."""
+    for col in data.columns:
+        if isinstance(data[col].iloc[0], str):
+            data[col] = data[col].str.wrap(wrap_len)
+
+
+def _get_tick_formatter() -> DatetimeTickFormatter:
+    """Return tick formatting for different zoom levels."""
+    # '%H:%M:%S.%3Nms
+    tick_format = DatetimeTickFormatter()
+    tick_format.days = ["%m-%d %H:%M"]
+    tick_format.hours = ["%H:%M:%S"]
+    tick_format.minutes = ["%H:%M:%S"]
+    tick_format.seconds = ["%H:%M:%S"]
+    tick_format.milliseconds = ["%H:%M:%S.%3N"]
+    return tick_format
+
+
+def _calc_auto_plot_height(group_count):
+    """Dynamic calculation of plot height."""
+    ht_per_row = 40
+    if group_count > 15:
+        ht_per_row = 25
+    return max(ht_per_row * group_count, 300)
+
+
+# pylint: disable=too-many-arguments
+def _create_range_tool(
+    data, min_time, max_time, plot_range, width, time_column: str = None
+):
+    """Create plot bar to act as as range selector."""
+    ext_min = min_time - ((max_time - min_time) * 0.15)
+    ext_max = max_time + ((max_time - min_time) * 0.15)
+    rng_select = figure(
+        x_range=(ext_min, ext_max),
+        title="Range Selector",
+        plot_height=100,
```

Comment:
This is too small on a high res display.  Maybe if we set this height as a percentage of the the main timeline height.

Reply:
Make 20% of main chart size